# NLP Project

## Project 1 Goals :
- clean the text data [ capiltalization and lower case / remove any urls or special charachters /
remove pancituation ]
- remove stop words 
- toknization 
- stemming


### Cleaning data is crucial for NLP as it enhances model accuracy by removing noise, inconsistencies, and irrelevant information, enabling algorithms to focus on meaningful patterns and relationships within the text.

## Project 2 Goals :
- bag of words 
- word embeddings 
- use bert and evaluate the model

## Importing Libararies Needed

In [90]:
# for data reading and slicing
import numpy as np
import pandas as pd

# for string manipulate
import string
import re

# Import Natural language Toolkit for stopwords and tokization and stemming
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer

# for train and test split 
from sklearn.model_selection import train_test_split

# Import for BOW
from sklearn.feature_extraction.text import CountVectorizer

# Import for word embedding
import gensim
from gensim.models import Word2Vec

# also for word embedding but different libarary
from sklearn.feature_extraction.text import TfidfVectorizer

# for randomly selecting the data
import random

# for BERT Neural Networks inside it
import tensorflow as tf
from keras.models import Model
from keras.layers import Input, Dense
from transformers import BertTokenizer, TFBertModel , TFBertForSequenceClassification
from keras.utils import to_categorical

# for evaluating the model 
from sklearn.metrics import classification_report

In [2]:
# Download NLTK tokenizer data
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [3]:
# Download NLTK Stop Words data
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\User\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Reading and Analyzing and Understanding the data

In [4]:
# Read CSV 
df = pd.read_csv(r"C:\Users\User\Desktop\koleya\summer 2\SPRINTS  AI & ML program\Assigment\assigment 6\train.csv")

# look inside the first 10 rows of data
df.head(50)

id                                       comment_text  \
0   0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1   000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2   000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3   0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4   0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
5   00025465d4725e87  "\n\nCongratulations from me as well, use the ...   
6   0002bcb3da6cb337       COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK   
7   00031b1e95af7921  Your vandalism to the Matt Shirvington article...   
8   00037261f536c51d  Sorry if the word 'nonsense' was offensive to ...   
9   00040093b2687caa  alignment on this subject and which are contra...   
10  0005300084f90edc  "\nFair use rationale for Image:Wonju.jpg\n\nT...   
11  00054a5e18b50dd4  bbq \n\nbe a man and lets discuss it-maybe ove...   
12  0005c987bdfc9d4b  Hey... what is it..\n@ | talk .\nWhat is it......   
13  0006f16e4e9f292e  Before you start throwing accusations and warn...   
14  00070ef96486d6f9  Oh, and the girl above started her arguments w...   
15  00078f8ce7eb276d  "\n\nJuelz Santanas Age\n\nIn 2002, Juelz Sant...   
16  0007e25b2121310b  Bye! \n\nDon't look, come or think of comming ...   
17  000897889268bc93   REDIRECT Talk:Voydan Pop Georgiev- Chernodrinski   
18  0009801bd85e5806  The Mitsurugi point made no sense - why not ar...   
19  0009eaea3325de8c  Don't mean to bother you \n\nI see that you're...   
20  000b08c464718505  "\n\n Regarding your recent edits \n\nOnce aga...   
21  000bfd0867774845  "\nGood to know. About me, yeah, I'm studying ...   
22  000c0dfd995809fa  "\n\n Snowflakes are NOT always symmetrical! \...   
23  000c6a3f0cd3ba8e  "\n\n The Signpost: 24 September 2012 \n\n Rea...   
24  000cfee90f50d471  "\n\nRe-considering 1st paragraph edit?\nI don...   
25  000eefc67a2c930f  Radial symmetry \n\nSeveral now extinct lineag...   
26  000f35deef84dc4a  There's no need to apologize. A Wikipedia arti...   
27  000ffab30195c5e1  Yes, because the mother of the child in the ca...   
28  0010307a3a50a353  "\nOk. But it will take a bit of work but I ca...   
29  0010833a96e1f886  "== A barnstar for you! ==\n\n  The Real Life ...   
30  0011cc71398479c4  How could I post before the block expires?  Th...   
31  00128363e367d703  Not sure about a heading of 'Fight for Freedom...   
32  001325b8b20ea8aa  Praise \n\nlooked at this article about 6 mont...   
33  001363e1dbe91225  I was able to post the above list so quickly b...   
34  0013a8b1a5f26bcb  "\nWell, not ""before the process"" but ""befo...   
35  00148d055a169b93  "\n\nNot at all, you are making a straw man ar...   
36  00151a9f93c6b059  "\n\n ""Mainland Asia"" includes ""the lower b...   
37  0015f4aa35ebe9b5  pretty much everyone from warren county/surrou...   
38  00169857adbc989b  Hi Explicit, can you block O Fenian for edit-w...   
39  0016e01b742b8da3  Notability of Rurika Kasuga\nA tag has been pl...   
40  001735f961a23fc4  "\n Sure, but the lead must briefly summarize ...   
41  00173958f46763a2  TFD \n\nI think we just eced. I think we respo...   
42  001810bf8c45bf5f  You are gay or antisemmitian? \n\nArchangel WH...   
43  00190820581d90ce           FUCK YOUR FILTHY MOTHER IN THE ASS, DRY!   
44  001956c382006abd  I'm Sorry \n\nI'm sorry I screwed around with ...   
45  001b2dd65d9d925c  I don't believe the Lisak criticism present th...   
46  001c419c445b5a59  You had a point, and it's now ammended with ap...   
47  001c557175094f10  In other words, you're too lazy to actually po...   
48  001cadfd324f8087  "\nAs for your claims of ""stalking"", that is...   
49  001d874a4d3e8813  "::::Jmabel; in regards to predominant scholar...   

    toxic  severe_toxic  obscene  threat  insult  identity_hate  
0       0             0        0       0       0              0  
1       0             0        0       0       0            

In [5]:
# get shape of data
df.shape 

(159571, 8)

#### There is 159571 rows and 8 columns

### Information of each column

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             159571 non-null  object
 1   comment_text   159571 non-null  object
 2   toxic          159571 non-null  int64 
 3   severe_toxic   159571 non-null  int64 
 4   obscene        159571 non-null  int64 
 5   threat         159571 non-null  int64 
 6   insult         159571 non-null  int64 
 7   identity_hate  159571 non-null  int64 
dtypes: int64(6), object(2)
memory usage: 9.7+ MB


### checking for nulls

In [7]:
# get the number of nulls in each column
df.isna().sum()

id               0
comment_text     0
toxic            0
severe_toxic     0
obscene          0
threat           0
insult           0
identity_hate    0
dtype: int64

#### there is no nulls in the data

### Describtion of data

In [8]:
# get the describtion of statistics of the int columns
df.describe().T

count      mean       std  min  25%  50%  75%  max
toxic          159571.0  0.095844  0.294379  0.0  0.0  0.0  0.0  1.0
severe_toxic   159571.0  0.009996  0.099477  0.0  0.0  0.0  0.0  1.0
obscene        159571.0  0.052948  0.223931  0.0  0.0  0.0  0.0  1.0
threat         159571.0  0.002996  0.054650  0.0  0.0  0.0  0.0  1.0
insult         159571.0  0.049364  0.216627  0.0  0.0  0.0  0.0  1.0
identity_hate  159571.0  0.008805  0.093420  0.0  0.0  0.0  0.0  1.0

### Types of columns

In [9]:
# get types of each column
df.dtypes

id               object
comment_text     object
toxic             int64
severe_toxic      int64
obscene           int64
threat            int64
insult            int64
identity_hate     int64
dtype: object

#### All columns are int except ID and comment_text

### Drop Duplicates (if there is duplicates)

In [10]:
# drop identical rows
df.drop_duplicates(subset='comment_text',inplace=True) 

In [11]:
df.shape 

(159571, 8)

### See unique values in data

In [12]:
# get the number of unique value of each column
df.nunique()

id               159571
comment_text     159571
toxic                 2
severe_toxic          2
obscene               2
threat                2
insult                2
identity_hate         2
dtype: int64

### See what is the unique values in each int column

In [13]:
# see each unique value in toxic quantity of it
df['toxic'].value_counts()

0    144277
1     15294
Name: toxic, dtype: int64

In [14]:
# see each unique value in severe_toxic quantity of it
df['severe_toxic'].value_counts()

0    157976
1      1595
Name: severe_toxic, dtype: int64

In [15]:
# see each unique value in obscene quantity of it
df['obscene'].value_counts()

0    151122
1      8449
Name: obscene, dtype: int64

In [16]:
# see each unique value in threat quantity of it
df['threat'].value_counts()

0    159093
1       478
Name: threat, dtype: int64

In [17]:
# see each unique value in insult quantity of it
df['insult'].value_counts()

0    151694
1      7877
Name: insult, dtype: int64

In [18]:
# see each unique value in identity_hate quantity of it
df['identity_hate'].value_counts()

0    158166
1      1405
Name: identity_hate, dtype: int64

#### I can say from in the above cells that all rows has unique id and comment_text , and the rest columns have only zero or one

### see if is there rows that has inappropriate comment and how many of the data it is

In [19]:
# get the rows that is appropriate by making a condition that all int values = 0
condition = (df["toxic"]==0)&(df["severe_toxic"]==0)&(df["obscene"]==0)&(df["threat"]==0)&(df["insult"]==0)&(df["identity_hate"]==0)

df[condition]

id                                       comment_text  \
0       0000997932d777bf  Explanation\nWhy the edits made under my usern...   
1       000103f0d9cfb60f  D'aww! He matches this background colour I'm s...   
2       000113f07ec002fd  Hey man, I'm really not trying to edit war. It...   
3       0001b41b1c6bb37e  "\nMore\nI can't make any real suggestions on ...   
4       0001d958c54c6e35  You, sir, are my hero. Any chance you remember...   
...                  ...                                                ...   
159566  ffe987279560d7ff  ":::::And for the second time of asking, when ...   
159567  ffea4adeee384e90  You should be ashamed of yourself \n\nThat is ...   
159568  ffee36eab5c267c9  Spitzer \n\nUmm, theres no actual article for ...   
159569  fff125370e4aaaf3  And it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  "\nAnd ... I really don't think you understand...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[143346 rows x 8 columns]

In [20]:
# get the number of rows that meets the condition
num = df[condition].shape[0] 

# divide the number by whole length of data * 100 to get percentage
Percentage_of_clean_data = (num / len(df))*100 

# get the percentage of appropriate data
Percentage_of_toxic_data = 100 - Percentage_of_clean_data

print("Percentage of clean data {} % \nPercentage of toxic data {} %".format(Percentage_of_clean_data,Percentage_of_toxic_data))

Percentage of clean data 89.83211235124176 % 
Percentage of toxic data 10.167887648758239 %


## Cleaning The Data

### lowercase the data

In [21]:
# make all data in comment_text column to lower case
df['comment_text'] = df['comment_text'].str.lower()

In [22]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation\nwhy the edits made under my usern...   
1       000103f0d9cfb60f  d'aww! he matches this background colour i'm s...   
2       000113f07ec002fd  hey man, i'm really not trying to edit war. it...   
3       0001b41b1c6bb37e  "\nmore\ni can't make any real suggestions on ...   
4       0001d958c54c6e35  you, sir, are my hero. any chance you remember...   
...                  ...                                                ...   
159566  ffe987279560d7ff  ":::::and for the second time of asking, when ...   
159567  ffea4adeee384e90  you should be ashamed of yourself \n\nthat is ...   
159568  ffee36eab5c267c9  spitzer \n\numm, theres no actual article for ...   
159569  fff125370e4aaaf3  and it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  "\nand ... i really don't think you understand...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

### Removing URLs

In [23]:
# function that removes urls from data
def remove_urls(dataframe, column_name):
    # put the style of url 
    url_pattern = r'https?://\S+|www\.\S+'
    
    # Function inside that removes URLs and replace it with space 
    def remove_url(text):
        return re.sub(url_pattern, ' ', text)
    
    # Apply the remove_url function to the column and moves on all rows
    dataframe[column_name] = dataframe[column_name].apply(lambda x: remove_url(x))
    
    return dataframe

#### count number of rows with URLS

In [24]:
url_pattern = r'https?://\S+|www\.\S+'

# Count rows with URLs 
num_rows_with_urls = df['comment_text'].str.contains(url_pattern, case=False, regex=True).sum()

num_rows_with_urls

5114

In [25]:
# apply function on data with mentioning the column
df = remove_urls(df, 'comment_text')

#### count number of rows with URLS after the function

In [26]:
# Count rows with URLs 
num_rows_with_urls = df['comment_text'].str.contains(url_pattern, case=False, regex=True).sum()

num_rows_with_urls

0

In [27]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation\nwhy the edits made under my usern...   
1       000103f0d9cfb60f  d'aww! he matches this background colour i'm s...   
2       000113f07ec002fd  hey man, i'm really not trying to edit war. it...   
3       0001b41b1c6bb37e  "\nmore\ni can't make any real suggestions on ...   
4       0001d958c54c6e35  you, sir, are my hero. any chance you remember...   
...                  ...                                                ...   
159566  ffe987279560d7ff  ":::::and for the second time of asking, when ...   
159567  ffea4adeee384e90  you should be ashamed of yourself \n\nthat is ...   
159568  ffee36eab5c267c9  spitzer \n\numm, theres no actual article for ...   
159569  fff125370e4aaaf3  and it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  "\nand ... i really don't think you understand...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

### Return words to it's basic form

In [28]:
# manual function that return Short Expression to it's basic form and also common punctuation in data and replace it with space
def remove(text):
    text = re.sub(r"what's", "what is ", text)
    text = re.sub(r"\'s", " ", text)
    text = re.sub(r"\'ve", " have ", text)
    text = re.sub(r"can't", "cannot ", text)
    text = re.sub(r"n't", " not ", text)
    text = re.sub(r"i'm", "i am ", text)
    text = re.sub(r"\'re", " are ", text)
    text = re.sub(r"\'d", " would ", text)
    text = re.sub(r"\'ll", " will ", text)
    text = re.sub(r"\'scuse", " excuse ", text)
    text = re.sub('\W', ' ', text)
    text = re.sub('\s+', ' ', text)
    text = re.sub('\n', '  ', text)
    text = text.strip(' ')
    return text

In [29]:
# apply function on mentioned column 
df['comment_text'] = df['comment_text'].map(lambda x : remove(x))

In [30]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation why the edits made under my userna...   
1       000103f0d9cfb60f  d aww he matches this background colour i am s...   
2       000113f07ec002fd  hey man i am really not trying to edit war it ...   
3       0001b41b1c6bb37e  more i cannot make any real suggestions on imp...   
4       0001d958c54c6e35  you sir are my hero any chance you remember wh...   
...                  ...                                                ...   
159566  ffe987279560d7ff  and for the second time of asking when your vi...   
159567  ffea4adeee384e90  you should be ashamed of yourself that is a ho...   
159568  ffee36eab5c267c9  spitzer umm theres no actual article for prost...   
159569  fff125370e4aaaf3  and it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  and i really do not think you understand i cam...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

### Remove special characters and numbers

In [31]:
# Function to remove special characters
def clean_text(text):
    return re.sub(r'[^A-Za-z\s]+', '', text)

#### count rows that have special characters and number

In [32]:
#the special characters
special_char = r'[^A-Za-z\s]+'

# function that counts special characters
def count_special_characters(text):
    return bool(re.search(special_char, text))

# count number of rows with special characters
rows_special_characters = df['comment_text'].apply(count_special_characters).sum()
rows_special_characters

53217

In [33]:
# apply function on mentioned column 
df['comment_text'] = df['comment_text'].apply(clean_text)

#### count rows that have special characters and numbers after the function

In [34]:
# count number of rows with special characters
rows_special_characters = df['comment_text'].apply(count_special_characters).sum()
rows_special_characters

0

In [35]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation why the edits made under my userna...   
1       000103f0d9cfb60f  d aww he matches this background colour i am s...   
2       000113f07ec002fd  hey man i am really not trying to edit war it ...   
3       0001b41b1c6bb37e  more i cannot make any real suggestions on imp...   
4       0001d958c54c6e35  you sir are my hero any chance you remember wh...   
...                  ...                                                ...   
159566  ffe987279560d7ff  and for the second time of asking when your vi...   
159567  ffea4adeee384e90  you should be ashamed of yourself that is a ho...   
159568  ffee36eab5c267c9  spitzer umm theres no actual article for prost...   
159569  fff125370e4aaaf3  and it looks like it was actually you who put ...   
159570  fff46fc426af1f9a  and i really do not think you understand i cam...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  
0           0             0        0       0       0              0  
1           0             0        0       0       0              0  
2           0             0        0       0       0              0  
3           0             0        0       0       0              0  
4           0             0        0       0       0              0  
...       ...           ...      ...     ...     ...            ...  
159566      0             0        0       0       0              0  
159567      0             0        0       0       0              0  
159568      0             0        0       0       0              0  
159569      0             0        0       0       0              0  
159570      0             0        0       0       0              0  

[159571 rows x 8 columns]

### Remove punctuation 

In [36]:
# function that removes punctuation and replace it with space
def remove_punctuation(text):
    
    # replace punctuation with space
    translator = str.maketrans('', '', string.punctuation)
    
    # clean data from punctuation
    cleaned_text = text.translate(translator)
    
    return cleaned_text

In [37]:
# apply function on mentioned column 
df['comment_text'] = df['comment_text'].map(lambda x : remove_punctuation(x))

In [38]:
df.head(10)

id                                       comment_text  toxic  \
0  0000997932d777bf  explanation why the edits made under my userna...      0   
1  000103f0d9cfb60f  d aww he matches this background colour i am s...      0   
2  000113f07ec002fd  hey man i am really not trying to edit war it ...      0   
3  0001b41b1c6bb37e  more i cannot make any real suggestions on imp...      0   
4  0001d958c54c6e35  you sir are my hero any chance you remember wh...      0   
5  00025465d4725e87  congratulations from me as well use the tools ...      0   
6  0002bcb3da6cb337       cocksucker before you piss around on my work      1   
7  00031b1e95af7921  your vandalism to the matt shirvington article...      0   
8  00037261f536c51d  sorry if the word nonsense was offensive to yo...      0   
9  00040093b2687caa  alignment on this subject and which are contra...      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0  
3             0        0       0       0              0  
4             0        0       0       0              0  
5             0        0       0       0              0  
6             1        1       0       1              0  
7             0        0       0       0              0  
8             0        0       0       0              0  
9             0        0       0       0              0

## Remove stop words

In [39]:
# check all stop words that will be removed in english language
stop_words = stopwords.words('english')
stop_words

['i',
 'me',
 'my',
 'myself',
 'we',
 'our',
 'ours',
 'ourselves',
 'you',
 "you're",
 "you've",
 "you'll",
 "you'd",
 'your',
 'yours',
 'yourself',
 'yourselves',
 'he',
 'him',
 'his',
 'himself',
 'she',
 "she's",
 'her',
 'hers',
 'herself',
 'it',
 "it's",
 'its',
 'itself',
 'they',
 'them',
 'their',
 'theirs',
 'themselves',
 'what',
 'which',
 'who',
 'whom',
 'this',
 'that',
 "that'll",
 'these',
 'those',
 'am',
 'is',
 'are',
 'was',
 'were',
 'be',
 'been',
 'being',
 'have',
 'has',
 'had',
 'having',
 'do',
 'does',
 'did',
 'doing',
 'a',
 'an',
 'the',
 'and',
 'but',
 'if',
 'or',
 'because',
 'as',
 'until',
 'while',
 'of',
 'at',
 'by',
 'for',
 'with',
 'about',
 'against',
 'between',
 'into',
 'through',
 'during',
 'before',
 'after',
 'above',
 'below',
 'to',
 'from',
 'up',
 'down',
 'in',
 'out',
 'on',
 'off',
 'over',
 'under',
 'again',
 'further',
 'then',
 'once',
 'here',
 'there',
 'when',
 'where',
 'why',
 'how',
 'all',
 'any',
 'both',
 'each

In [40]:
# Function to remove stop words
def remove_stop_words_from_column(text):  
    # tokenizes the  text  into individual words using the  word_tokenize  function. 
    word_tokens = word_tokenize(text)
    
    # collect all words not included in stop_word list by iterating over each word in word_tokens and put it in filtered_text
    filtered_text = [word for word in word_tokens if word not in stop_words]
    
    # Return the filtered text
    return ' '.join(filtered_text)

#### count stop words before the function

In [41]:
# Collect all stop words and save it to variable (stop_words)
stop_words = set(stopwords.words('english'))

# Function to check for number of stop words
def count_rows_stop_words(text):
    tokens = nltk.word_tokenize(text)
    return any(token in stop_words for token in tokens)

# Count the number of rows with stop words
rows_with_stop_words = df['comment_text'].apply(count_rows_stop_words).sum()
rows_with_stop_words

155732

In [42]:
# apply function on mentioned column 
df['comment_text'] = df['comment_text'].apply(remove_stop_words_from_column)

#### count stop words after the function

In [43]:
rows_with_stop_words = df['comment_text'].apply(count_rows_stop_words).sum()
rows_with_stop_words

0

In [44]:
df.head(10)

id                                       comment_text  toxic  \
0  0000997932d777bf  explanation edits made username hardcore metal...      0   
1  000103f0d9cfb60f  aww matches background colour seemingly stuck ...      0   
2  000113f07ec002fd  hey man really trying edit war guy constantly ...      0   
3  0001b41b1c6bb37e  make real suggestions improvement wondered sec...      0   
4  0001d958c54c6e35                      sir hero chance remember page      0   
5  00025465d4725e87           congratulations well use tools well talk      0   
6  0002bcb3da6cb337                        cocksucker piss around work      1   
7  00031b1e95af7921  vandalism matt shirvington article reverted pl...      0   
8  00037261f536c51d  sorry word nonsense offensive anyway intending...      0   
9  00040093b2687caa               alignment subject contrary dulithgow      0   

   severe_toxic  obscene  threat  insult  identity_hate  
0             0        0       0       0              0  
1             0        0       0       0              0  
2             0        0       0       0              0  
3             0        0       0       0              0  
4             0        0       0       0              0  
5             0        0       0       0              0  
6             1        1       0       1              0  
7             0        0       0       0              0  
8             0        0       0       0              0  
9             0        0       0       0              0

## Toknization

In [45]:
# function that tokenize the data which is built in 
def tokenize_text(text):
    tokens = word_tokenize(text)
    return tokens

In [46]:
# apply function on mentioned column 
df['tokenized_column'] = df['comment_text'].apply(tokenize_text)

In [47]:
df.head(10)

id                                       comment_text  toxic  \
0  0000997932d777bf  explanation edits made username hardcore metal...      0   
1  000103f0d9cfb60f  aww matches background colour seemingly stuck ...      0   
2  000113f07ec002fd  hey man really trying edit war guy constantly ...      0   
3  0001b41b1c6bb37e  make real suggestions improvement wondered sec...      0   
4  0001d958c54c6e35                      sir hero chance remember page      0   
5  00025465d4725e87           congratulations well use tools well talk      0   
6  0002bcb3da6cb337                        cocksucker piss around work      1   
7  00031b1e95af7921  vandalism matt shirvington article reverted pl...      0   
8  00037261f536c51d  sorry word nonsense offensive anyway intending...      0   
9  00040093b2687caa               alignment subject contrary dulithgow      0   

   severe_toxic  obscene  threat  insult  identity_hate  \
0             0        0       0       0              0   
1             0        0       0       0              0   
2             0        0       0       0              0   
3             0        0       0       0              0   
4             0        0       0       0              0   
5             0        0       0       0              0   
6             1        1       0       1              0   
7             0        0       0       0              0   
8             0        0       0       0              0   
9             0        0       0       0              0   

                                    tokenized_column  
0  [explanation, edits, made, username, hardcore,...  
1  [aww, matches, background, colour, seemingly, ...  
2  [hey, man, really, trying, edit, war, guy, con...  
3  [make, real, suggestions, improvement, wondere...  
4                [sir, hero, chance, remember, page]  
5    [congratulations, well, use, tools, well, talk]  
6                   [cocksucker, piss, around, work]  
7  [vandalism, matt, shirvington, article, revert...  
8  [sorry, word, nonsense, offensive, anyway, int...  
9          [alignment, subject, contrary, dulithgow]

## Stemming

In [48]:
# Initialize the SnowballStemmer
stemmer = SnowballStemmer('english')

# Apply stemming to an already tokenized column
def stem_tokens(tokens):
    stemmed_tokens = [stemmer.stem(token) for token in tokens]
    return stemmed_tokens

In [49]:
# apply function on mentioned column 
df['stemmed_column'] = df['tokenized_column'].apply(stem_tokens)

In [50]:
df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  aww matches background colour seemingly stuck ...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
159566  ffe987279560d7ff  second time asking view completely contradicts...   
159567  ffea4adeee384e90               ashamed horrible thing put talk page   
159568  ffee36eab5c267c9  spitzer umm theres actual article prostitution...   
159569  fff125370e4aaaf3  looks like actually put speedy first version d...   
159570  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
159566      0             0        0       0       0              0   
159567      0             0        0       0       0              0   
159568      0             0        0       0       0              0   
159569      0             0        0       0       0              0   
159570      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [aww, matches, background, colour, seemingly, ...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
159566  [second, time, asking, view, completely, contr...   
159567        [ashamed, horrible, thing, put, talk, page]   
159568  [spitzer, umm, theres, actual, article, prosti...   
159569  [looks, like, actually, put, speedy, first, ve...   
159570  [really, think, understand, came, idea, bad, r...   

                                           stemmed_column  
0       [explan, edit, made, usernam, hardcor, metalli...  
1       [aww, match, background, colour, seem, stuck, ...  
2       [hey, man, realli, tri, edit, war, guy, consta...  
3       [make, real, suggest, improv, wonder, section,...  
4                        [sir, hero, chanc, rememb, page]  
...                                                   ...  
159566  [second, time, ask, view, complet, contradict,...  
159567           [asham, horribl, thing, put, talk, page]  
159568  [spitzer, umm, there, actual, articl, prostitu...  
159569  [look, like, actual, put, speedi, first, versi...  
159570  [realli, think, understand, came, idea, bad, r...  

[159571 rows x 10 columns]

## Split appropriate and inappropriate data

In [51]:
# put the data that meet the condition made above 
appropriate_data = df[condition]

# reset the index
appropriate_data.reset_index(drop=True, inplace=True)

appropriate_data

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  aww matches background colour seemingly stuck ...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
143341  ffe987279560d7ff  second time asking view completely contradicts...   
143342  ffea4adeee384e90               ashamed horrible thing put talk page   
143343  ffee36eab5c267c9  spitzer umm theres actual article prostitution...   
143344  fff125370e4aaaf3  looks like actually put speedy first version d...   
143345  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
143341      0             0        0       0       0              0   
143342      0             0        0       0       0              0   
143343      0             0        0       0       0              0   
143344      0             0        0       0       0              0   
143345      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [aww, matches, background, colour, seemingly, ...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
143341  [second, time, asking, view, completely, contr...   
143342        [ashamed, horrible, thing, put, talk, page]   
143343  [spitzer, umm, theres, actual, article, prosti...   
143344  [looks, like, actually, put, speedy, first, ve...   
143345  [really, think, understand, came, idea, bad, r...   

                                           stemmed_column  
0       [explan, edit, made, usernam, hardcor, metalli...  
1       [aww, match, background, colour, seem, stuck, ...  
2       [hey, man, realli, tri, edit, war, guy, consta...  
3       [make, real, suggest, improv, wonder, section,...  
4                        [sir, hero, chanc, rememb, page]  
...                                                   ...  
143341  [second, time, ask, view, complet, contradict,...  
143342           [asham, horribl, thing, put, talk, page]  
143343  [spitzer, umm, there, actual, articl, prostitu...  
143344  [look, like, actual, put, speedi, first, versi...  
143345  [realli, think, understand, came, idea, bad, r...  

[143346 rows x 10 columns]

In [52]:
# put the data that doesnot meet the condition made above 
inappropriate_data = df[~condition]

# reset the index
inappropriate_data.reset_index(drop=True, inplace=True)

inappropriate_data

id                                       comment_text  \
0      0002bcb3da6cb337                        cocksucker piss around work   
1      0005c987bdfc9d4b  hey talk exclusive group wp talibans good dest...   
2      0007e25b2121310b            bye look come think comming back tosser   
3      001810bf8c45bf5f  gay antisemmitian archangel white tiger meow g...   
4      00190820581d90ce                         fuck filthy mother ass dry   
...                 ...                                                ...   
16220  fef4cf7ba0012866  previous conversation fucking shit eating libe...   
16221  ff39a2895fc3b40e                            mischievious pubic hair   
16222  ffa33d3122b599d6  absurd edits absurd edits great white shark to...   
16223  ffb47123b2d82762  hey listen ever delete edits ever annoyed wwe ...   
16224  ffbdbb0483ed0841  going keep posting stuff u deleted fucking sit...   

       toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0          1             1        1       0       1              0   
1          1             0        0       0       0              0   
2          1             0        0       0       0              0   
3          1             0        1       0       1              1   
4          1             0        1       0       1              0   
...      ...           ...      ...     ...     ...            ...   
16220      1             0        1       0       1              1   
16221      1             0        0       0       1              0   
16222      1             0        1       0       1              0   
16223      1             0        0       0       1              0   
16224      1             0        1       0       1              0   

                                        tokenized_column  \
0                       [cocksucker, piss, around, work]   
1      [hey, talk, exclusive, group, wp, talibans, go...   
2        [bye, look, come, think, comming, back, tosser]   
3      [gay, antisemmitian, archangel, white, tiger, ...   
4                       [fuck, filthy, mother, ass, dry]   
...                                                  ...   
16220  [previous, conversation, fucking, shit, eating...   
16221                        [mischievious, pubic, hair]   
16222  [absurd, edits, absurd, edits, great, white, s...   
16223  [hey, listen, ever, delete, edits, ever, annoy...   
16224  [going, keep, posting, stuff, u, deleted, fuck...   

                                          stemmed_column  
0                         [cocksuck, piss, around, work]  
1      [hey, talk, exclus, group, wp, taliban, good, ...  
2            [bye, look, come, think, com, back, tosser]  
3      [gay, antisemmitian, archangel, white, tiger, ...  
4                       [fuck, filthi, mother, ass, dri]  
...                                                  ...  
16220  [previous, convers, fuck, shit, eat, liber, ma...  
16221                           [mischievi, pubic, hair]  
16222  [absurd, edit, absurd, edit, great, white, sha...  
16223  [hey, listen, ever, delet, edit, ever, annoy, ...  
16224  [go, keep, post, stuff, u, delet, fuck, site, ...  

[16225 rows x 10 columns]

## Count the occurrences of each word

In [53]:
# separate stemmed column from an array to be a sentence , this will help in further models
df['stemmed_new_column'] = df['stemmed_column'].apply(lambda arr: ' '.join(arr))

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  aww matches background colour seemingly stuck ...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
159566  ffe987279560d7ff  second time asking view completely contradicts...   
159567  ffea4adeee384e90               ashamed horrible thing put talk page   
159568  ffee36eab5c267c9  spitzer umm theres actual article prostitution...   
159569  fff125370e4aaaf3  looks like actually put speedy first version d...   
159570  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
159566      0             0        0       0       0              0   
159567      0             0        0       0       0              0   
159568      0             0        0       0       0              0   
159569      0             0        0       0       0              0   
159570      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [aww, matches, background, colour, seemingly, ...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
159566  [second, time, asking, view, completely, contr...   
159567        [ashamed, horrible, thing, put, talk, page]   
159568  [spitzer, umm, theres, actual, article, prosti...   
159569  [looks, like, actually, put, speedy, first, ve...   
159570  [really, think, understand, came, idea, bad, r...   

                                           stemmed_column  \
0       [explan, edit, made, usernam, hardcor, metalli...   
1       [aww, match, background, colour, seem, stuck, ...   
2       [hey, man, realli, tri, edit, war, guy, consta...   
3       [make, real, suggest, improv, wonder, section,...   
4                        [sir, hero, chanc, rememb, page]   
...                                                   ...   
159566  [second, time, ask, view, complet, contradict,...   
159567           [asham, horribl, thing, put, talk, page]   
159568  [spitzer, umm, there, actual, articl, prostitu...   
159569  [look, like, actual, put, speedi, first, versi...   
159570  [realli, think, understand, came, idea, bad, r...   

                                       stemmed_new_column  
0       explan edit made usernam hardcor metallica fan...  
1       aww match background colour seem stuck thank t...  
2       hey man realli tri edit war guy constant remov...  
3       make real suggest improv wonder section statis...  
4                              sir hero chanc rememb page  
...                                                   ...  
159566  second time ask view complet contradict covera...  
159567                  asham horribl thing put talk page  
159568  spitzer umm there actual articl prostitut ring...  
159569  look like actual put speedi first version dele...  
159570  realli think understand came idea bad right aw...  

[159571 rows x 11 columns]

In [54]:
# initialize a dictionary of words
count_occurrences = {}

# iterate over each row in data
for index, row in df.iterrows():

    # split each sentence to words
    tokens = row['stemmed_new_column'].split()  

    # for each word in text if appeared before add one to it  
    for token in tokens:
        if token in count_occurrences:
            count_occurrences[token] += 1

            # if appeared for the first time initialize it with one
        else:
            count_occurrences[token] = 1

In [55]:
count_occurrences

{'explan': 2073,
 'edit': 41538,
 'made': 9681,
 'usernam': 1929,
 'hardcor': 166,
 'metallica': 38,
 'fan': 1598,
 'revert': 10293,
 'vandal': 10609,
 'closur': 97,
 'gas': 329,
 'vote': 2308,
 'new': 10651,
 'york': 929,
 'doll': 39,
 'fac': 352,
 'pleas': 29971,
 'remov': 15596,
 'templat': 4638,
 'talk': 39528,
 'page': 57087,
 'sinc': 8693,
 'retir': 446,
 'aww': 21,
 'match': 1076,
 'background': 2289,
 'colour': 365,
 'seem': 10597,
 'stuck': 339,
 'thank': 23990,
 'januari': 1171,
 'utc': 6878,
 'hey': 3193,
 'man': 3179,
 'realli': 9209,
 'tri': 11549,
 'war': 6084,
 'guy': 3723,
 'constant': 778,
 'relev': 2893,
 'inform': 13646,
 'instead': 3617,
 'care': 4593,
 'format': 1706,
 'actual': 7732,
 'info': 2650,
 'make': 20878,
 'real': 3846,
 'suggest': 6326,
 'improv': 3608,
 'wonder': 2225,
 'section': 11625,
 'statist': 586,
 'later': 2460,
 'subsect': 251,
 'type': 2886,
 'accid': 312,
 'think': 21922,
 'refer': 12931,
 'may': 15619,
 'need': 15951,
 'tidi': 124,
 'exact':

In [56]:
# convert dictionary to dataframe
count_occurrences_df = pd.DataFrame(list(count_occurrences.items()), columns=['Word', 'Occurrences'], index=range(1, len(count_occurrences) + 1))

count_occurrences_df

Word  Occurrences
1            explan         2073
2              edit        41538
3              made         9681
4           usernam         1929
5           hardcor          166
...             ...          ...
129756   webaddress            1
129757      gratest            1
129758  hanumakonda            1
129759      automak            1
129760          ciu            1

[129760 rows x 2 columns]

In [57]:
# get index of the largest number in Occurrences
max =count_occurrences_df['Occurrences'].argmax()

# print the word that appeared the most
count_occurrences_df['Word'][max]

'backlog'

### word "backlog" is the most appeared word in the data

### after tokenization to make BOW we need to count occurrences of each word so here is a manual dictionary does this

## Bag Of Words (BOW)

In [58]:
# Initialize the CountVectorizer
vectorizer = CountVectorizer(max_features=5000)
# make max features = 5000 because if it left default it will take all features and will make MEMORY ERROR

In [59]:
# Fit the stemmed column
bow_matrix = vectorizer.fit_transform(df['stemmed_new_column'])

In [60]:
# convert the matrix to dataframe 
bow_df=pd.DataFrame(bow_matrix.toarray(),columns=vectorizer.get_feature_names_out())

In [61]:
bow_df

aa  aaron  ab  abandon  abbrevi  abc  abid  abil  abl  abort  ...  \
0        0      0   0        0        0    0     0     0    0      0  ...   
1        0      0   0        0        0    0     0     0    0      0  ...   
2        0      0   0        0        0    0     0     0    0      0  ...   
3        0      0   0        0        0    0     0     0    0      0  ...   
4        0      0   0        0        0    0     0     0    0      0  ...   
...     ..    ...  ..      ...      ...  ...   ...   ...  ...    ...  ...   
159566   0      0   0        0        0    0     0     0    0      0  ...   
159567   0      0   0        0        0    0     0     0    0      0  ...   
159568   0      0   0        0        0    0     0     0    0      0  ...   
159569   0      0   0        0        0    0     0     0    0      0  ...   
159570   0      0   0        0        0    0     0     0    0      0  ...   

        yourselfgo  youth  youtub  ytmnd  yugoslav  yugoslavia  zealand  zero  \
0                0      0       0      0         0           0        0     0   
1                0      0       0      0         0           0        0     0   
2                0      0       0      0         0           0        0     0   
3                0      0       0      0         0           0        0     0   
4                0      0       0      0         0           0        0     0   
...            ...    ...     ...    ...       ...         ...      ...   ...   
159566           0      0       0      0         0           0        0     0   
159567           0      0       0      0         0           0        0     0   
159568           0      0       0      0         0           0        0     0   
159569           0      0       0      0         0           0        0     0   
159570           0      0       0      0         0           0        0     0   

        zionist  zone  
0             0     0  
1             0     0  
2             0     0  
3             0     0  
4             0     0  
...         ...   ...  
159566        0     0  
159567        0     0  
159568        0     0  
159569        0     0  
159570        0     0  

[159571 rows x 5000 columns]

### the columns the number each word and rows is each sentence and the score is if the word in the sentence and how many words in there

### (If we made max features default which takes all features) we cannot convert the bow_matrix to dataframe nor array because data is to large to convert and will get memory error

### Another way to display by printing the sparse matrix

In [62]:
# Initialize the CountVectorizer
vectorizer = CountVectorizer()

# Fit the stemmed column
bow_matrix = vectorizer.fit_transform(df['stemmed_new_column'])

In [63]:
print(bow_matrix)

  (0, 37299)	1
  (0, 33416)	1
  (0, 67489)	1
  (0, 120056)	1
  (0, 47989)	1
  (0, 71495)	1
  (0, 38070)	1
  (0, 95398)	1
  (0, 120658)	1
  (0, 21455)	1
  (0, 42819)	1
  (0, 122566)	1
  (0, 77621)	1
  (0, 128364)	1
  (0, 31157)	1
  (0, 37625)	1
  (0, 87362)	1
  (0, 94555)	1
  (0, 112269)	1
  (0, 111101)	1
  (0, 83345)	1
  (0, 103507)	1
  (0, 95267)	1
  (1, 111101)	1
  (1, 8665)	1
  :	:
  (159568, 25431)	1
  (159568, 106460)	1
  (159569, 1020)	1
  (159569, 39655)	1
  (159569, 28025)	1
  (159569, 106223)	1
  (159569, 66257)	2
  (159569, 65091)	1
  (159569, 121488)	1
  (159569, 91107)	1
  (159570, 93275)	1
  (159570, 54580)	1
  (159570, 113566)	1
  (159570, 44740)	1
  (159570, 17097)	1
  (159570, 9118)	2
  (159570, 118452)	1
  (159570, 49093)	1
  (159570, 44861)	1
  (159570, 95918)	1
  (159570, 60784)	1
  (159570, 22511)	1
  (159570, 8607)	2
  (159570, 52421)	2
  (159570, 95485)	1


#### it only display the words index that appear in the text and how many times it appear , any other word that isn't in data donn't appear

#### cannot convert to Dataframe nor to array because it will get memory error

## Word Embedding

### Using TF-IDF libarary

In [ ]:
# Initialize the TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=5000)
# make max features = 5000 because if it left default it will take all features and will make MEMORY ERROR

In [ ]:
# Fit on stemmed column
tfidf_matrix = tfidf_vectorizer.fit_transform(df['stemmed_new_column'])

In [ ]:
# Get the names from the Vectorizer
feature_names = tfidf_vectorizer.get_feature_names_out()

In [85]:
# convert the matrix to dataframe 
tf_idf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=feature_names)

tf_idf_df

aa  aaron   ab  abandon  abbrevi  abc  abid  abil  abl  abort  ...  \
0       0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
1       0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
2       0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
3       0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
4       0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
...     ...    ...  ...      ...      ...  ...   ...   ...  ...    ...  ...   
159566  0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
159567  0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
159568  0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
159569  0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   
159570  0.0    0.0  0.0      0.0      0.0  0.0   0.0   0.0  0.0    0.0  ...   

        yourselfgo  youth  youtub  ytmnd  yugoslav  yugoslavia  zealand  zero  \
0              0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
1              0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
2              0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
3              0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
4              0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
...            ...    ...     ...    ...       ...         ...      ...   ...   
159566         0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
159567         0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
159568         0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
159569         0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   
159570         0.0    0.0     0.0    0.0       0.0         0.0      0.0   0.0   

        zionist  zone  
0           0.0   0.0  
1           0.0   0.0  
2           0.0   0.0  
3           0.0   0.0  
4           0.0   0.0  
...         ...   ...  
159566      0.0   0.0  
159567      0.0   0.0  
159568      0.0   0.0  
159569      0.0   0.0  
159570      0.0   0.0  

[159571 rows x 5000 columns]

##### columns is each word in data and rows is each sentence in data and the values inside is TF-IDF Score: obtained by multiplying the Term Frequency (TF) and Inverse Document Frequency (IDF)

##### (If we made max features default which takes all features) we cannot convert the tfidf_matrix to dataframe nor array because data is to large to convert and will get memory error

#### Another way to display by printing the sparse matrix

In [86]:
# Initialize the TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer()

In [87]:
# Fit on stemmed column
tfidf_matrix = tfidf_vectorizer.fit_transform(df['stemmed_new_column'])

In [89]:
print(tfidf_matrix)

  (0, 95267)	0.24876749235093354
  (0, 103507)	0.1425103519507942
  (0, 83345)	0.09187496318857917
  (0, 111101)	0.09544168658706023
  (0, 112269)	0.17521790623396125
  (0, 94555)	0.12580694913164783
  (0, 87362)	0.10402557628201634
  (0, 37625)	0.2584408804975239
  (0, 31157)	0.3326925689233515
  (0, 128364)	0.22578870561478556
  (0, 77621)	0.14014776110878466
  (0, 122566)	0.1980835265055874
  (0, 42819)	0.26744044548921564
  (0, 21455)	0.2982847822899369
  (0, 120658)	0.14124111094838715
  (0, 95398)	0.14272721783466202
  (0, 38070)	0.20987332895643165
  (0, 71495)	0.33845636962025183
  (0, 47989)	0.29486886718636757
  (0, 120056)	0.20401055464181858
  (0, 67489)	0.1393871474195857
  (0, 33416)	0.10062462740915695
  (0, 37299)	0.19501396031651005
  (1, 120212)	0.2206220641200648
  (1, 56614)	0.31651722192919096
  :	:
  (159568, 6855)	0.10236704952781132
  (159568, 1020)	0.17508993548007185
  (159569, 91107)	0.31763069313005987
  (159569, 121488)	0.3794404137725878
  (159569, 65091)	

##### it only display the words index that appear in the text and TF-IDF score of it , any other word that isn't in data donn't appear

##### cannot convert to Dataframe nor to array because it will get memory error

### Using Word2Vec libarary

In [65]:
# Assuming you have a list of stemmed sentences
stemmed_sentences = df['stemmed_new_column']

# split each sentence to words
sentence_words = [sentence.split() for sentence in stemmed_sentences]

# Train the model
# best range of vector size is between (100-300) which is dimension
# window is distance between current and predicted word 
# workers is umber of CPU cores to use for training
model = Word2Vec(sentence_words, vector_size=300, window=5, min_count=1, workers=4)

# Get the word embeddings for all words
word_embeddings = model.wv

In [66]:
# Collect word embeddings in a list
embedding_list = []

# List to store the corresponding words
word_list = []

# iterate over each word in data
for word in word_embeddings.key_to_index:
    # put each score of word in a variable
    embedding = word_embeddings[word]

    # then append it in an array
    embedding_list.append(embedding)

    # apeend the word in an array to be the index
    word_list.append(word)

In [67]:
# Create a dataframe from the embedding list with words as index
word_embedding_df = pd.DataFrame(embedding_list, index=word_list)

word_embedding_df

0         1         2         3         4         5    \
articl      -0.033067  0.624858  0.067449  0.600456 -0.778151 -1.175725   
page        -0.639260  1.358475  0.623980 -0.674487 -0.276724 -0.588981   
wikipedia   -0.363596 -0.939207 -0.341324  0.466016  0.067741 -0.602594   
edit        -0.835617 -0.473956  0.285923  0.217409 -0.317279  0.020332   
talk        -1.616057 -0.415388  1.657600 -1.227174 -1.285565  0.840989   
...               ...       ...       ...       ...       ...       ...   
bloy        -0.009626  0.000788 -0.007648 -0.000331  0.005861 -0.009028   
chancelleri  0.016161 -0.001543 -0.005588  0.012852 -0.005206 -0.010158   
portant      0.000991  0.024425 -0.000748  0.014609 -0.001689 -0.018852   
ouranio      0.001870  0.020853  0.002058  0.010674 -0.011549 -0.013351   
ciu         -0.004756  0.006472  0.001600 -0.003440  0.000200 -0.010320   

                  6         7         8         9    ...       290       291  \
articl      -0.121369  0.590934  0.112086  0.405279  ... -0.700605  0.481617   
page         0.754116  0.374246  1.206737 -1.551515  ... -0.549755  0.348929   
wikipedia    0.002435 -0.081889 -0.187970 -1.068749  ... -1.880256 -0.224864   
edit         1.301405  0.218953  1.236998 -1.677666  ...  0.810417 -1.489304   
talk         0.403911  0.185184  1.242798 -0.359630  ... -0.064096 -0.129708   
...               ...       ...       ...       ...  ...       ...       ...   
bloy         0.010184  0.013564  0.001936 -0.009733  ...  0.010767  0.020708   
chancelleri  0.005121  0.020059  0.005167  0.007900  ... -0.009651  0.006780   
portant      0.021150  0.044257  0.001460  0.005466  ... -0.003194  0.019379   
ouranio      0.003710  0.030541  0.002877  0.001719  ...  0.003123  0.018091   
ciu          0.007788  0.034610  0.013540 -0.003301  ... -0.004210  0.005093   

                  292       293       294       295       296       297  \
articl       0.218966 -1.013990  0.349719  1.364781 -0.398659  0.060480   
page         0.061990 -0.886083  1.050348  0.725704 -0.681177 -0.718759   
wikipedia    1.115999 -0.638115  0.728477  0.325811 -0.424821  0.768273   
edit        -1.073203 -0.761311  0.765692  0.312403 -1.825217  0.843992   
talk         0.283196 -1.316172  0.739521  0.148025 -1.303362  0.473901   
...               ...       ...       ...       ...       ...       ...   
bloy        -0.008220  0.010558  0.004824  0.007446 -0.004059 -0.003896   
chancelleri -0.001200  0.009688  0.001127 -0.001003  0.020858 -0.010726   
portant      0.001028 -0.001787  0.017768  0.020610  0.003445 -0.017339   
ouranio      0.008880  0.003622  0.009793  0.016214 -0.005104 -0.015988   
ciu          0.015190  0.010523  0.019578  0.028828  0.005893 -0.008494   

                  298       299  
articl       0.684610 -0.586724  
page         0.656708 -0.603569  
wikipedia    0.670248  0.272175  
edit        -0.713610  0.827022  
talk        -0.702880  0.990956  
...               ...       ...  
bloy         0.007486  0.000484  
chancelleri  0.013726 -0.003206  
portant      0.013700 -0.011804  
ouranio      0.004342 -0.002982  
ciu          0.021362  0.007324  

[129760 rows x 300 columns]

#### dataframe has 300 columns which is the dimension and rows which is each word in text , and the scores is semantic and syntactic information about the word

## BERT

### new column is made containing binary if it has offensive comment or not

#### made offensive column because it is more practical to have binary classification model that tells you this is offensive or non-offensive comment for real life applications 

In [72]:
# Create a new column with make value equal 1
df['offensive'] = 1

# if it meet the condition and doesnot has offensive comment make the value equal 0
df.loc[condition, 'offensive'] = 0

df

id                                       comment_text  \
0       0000997932d777bf  explanation edits made username hardcore metal...   
1       000103f0d9cfb60f  aww matches background colour seemingly stuck ...   
2       000113f07ec002fd  hey man really trying edit war guy constantly ...   
3       0001b41b1c6bb37e  make real suggestions improvement wondered sec...   
4       0001d958c54c6e35                      sir hero chance remember page   
...                  ...                                                ...   
159566  ffe987279560d7ff  second time asking view completely contradicts...   
159567  ffea4adeee384e90               ashamed horrible thing put talk page   
159568  ffee36eab5c267c9  spitzer umm theres actual article prostitution...   
159569  fff125370e4aaaf3  looks like actually put speedy first version d...   
159570  fff46fc426af1f9a  really think understand came idea bad right aw...   

        toxic  severe_toxic  obscene  threat  insult  identity_hate  \
0           0             0        0       0       0              0   
1           0             0        0       0       0              0   
2           0             0        0       0       0              0   
3           0             0        0       0       0              0   
4           0             0        0       0       0              0   
...       ...           ...      ...     ...     ...            ...   
159566      0             0        0       0       0              0   
159567      0             0        0       0       0              0   
159568      0             0        0       0       0              0   
159569      0             0        0       0       0              0   
159570      0             0        0       0       0              0   

                                         tokenized_column  \
0       [explanation, edits, made, username, hardcore,...   
1       [aww, matches, background, colour, seemingly, ...   
2       [hey, man, really, trying, edit, war, guy, con...   
3       [make, real, suggestions, improvement, wondere...   
4                     [sir, hero, chance, remember, page]   
...                                                   ...   
159566  [second, time, asking, view, completely, contr...   
159567        [ashamed, horrible, thing, put, talk, page]   
159568  [spitzer, umm, theres, actual, article, prosti...   
159569  [looks, like, actually, put, speedy, first, ve...   
159570  [really, think, understand, came, idea, bad, r...   

                                           stemmed_column  \
0       [explan, edit, made, usernam, hardcor, metalli...   
1       [aww, match, background, colour, seem, stuck, ...   
2       [hey, man, realli, tri, edit, war, guy, consta...   
3       [make, real, suggest, improv, wonder, section,...   
4                        [sir, hero, chanc, rememb, page]   
...                                                   ...   
159566  [second, time, ask, view, complet, contradict,...   
159567           [asham, horribl, thing, put, talk, page]   
159568  [spitzer, umm, there, actual, articl, prostitu...   
159569  [look, like, actual, put, speedi, first, versi...   
159570  [realli, think, understand, came, idea, bad, r...   

                                       stemmed_new_column  offensive  
0       explan edit made usernam hardcor metallica fan...          0  
1       aww match background colour seem stuck thank t...          0  
2       hey man realli tri edit war guy constant remov...          0  
3       make real suggest improv wonder section statis...          0  
4                              sir hero chanc rememb page          0  
...                                                   ...        ...  
159566  second time ask view complet contradict covera...          0  
159567                  asham horribl thing put talk page          0  
159568  spitzer umm there actual articl prostitut ring...          0  
159569  look like actual put speedi first ve

### New dataframe is made to enter BERT model containing two columns the comments and offensive column

#### why the comment_text not stemmed_new_column Because the BERT model can leverage the full range of vocabulary and contextual information to make more accurate predictions and by using stemmed column may lead to loss of semantic information and context

In [73]:
# make new dataframe containing only needed columns for bert model
new_df = df.loc[:, ['comment_text','offensive']]

new_df

comment_text  offensive
0       explanation edits made username hardcore metal...          0
1       aww matches background colour seemingly stuck ...          0
2       hey man really trying edit war guy constantly ...          0
3       make real suggestions improvement wondered sec...          0
4                           sir hero chance remember page          0
...                                                   ...        ...
159566  second time asking view completely contradicts...          0
159567               ashamed horrible thing put talk page          0
159568  spitzer umm theres actual article prostitution...          0
159569  looks like actually put speedy first version d...          0
159570  really think understand came idea bad right aw...          0

[159571 rows x 2 columns]

### Remove some rows from data 

#### why remove some rows from data ? because if we used all the data the model will run in many hours and will exceed the deadline 

In [74]:
# will remove the 40 % of data 
percentage_to_remove = 0.4
rows_to_remove = int(len(new_df) * percentage_to_remove)

# Randomly select rows to remove
random.seed(42)  
rows_to_remove_indices = random.sample(range(len(new_df)), rows_to_remove)

# new dataframe by reseting the index
df_filtered = new_df.drop(rows_to_remove_indices).reset_index()

In [75]:
# drop column index
df_filtered.drop('index',axis=1,inplace=True)

In [76]:
df_filtered

comment_text  offensive
0      explanation edits made username hardcore metal...          0
1      aww matches background colour seemingly stuck ...          0
2      make real suggestions improvement wondered sec...          0
3                          sir hero chance remember page          0
4               congratulations well use tools well talk          0
...                                                  ...        ...
95738  catalan independentism social movement involvi...          0
95739  second time asking view completely contradicts...          0
95740  spitzer umm theres actual article prostitution...          0
95741  looks like actually put speedy first version d...          0
95742  really think understand came idea bad right aw...          0

[95743 rows x 2 columns]

### keras 

In [78]:
# Split the data into input text and labels of the column offensive and get value as numpy array
text = df_filtered['comment_text'].values
labels = df_filtered['offensive'].values

# Convert labels to binary matrix
labels = to_categorical(labels)


In [ ]:
# Split the data train and test 
train_text, test_text, train_labels, test_labels = train_test_split(text, labels, test_size=0.2, random_state=42)

In [79]:
# download pretrained model for BERT
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
bert_model = TFBertModel.from_pretrained('bert-base-uncased')

Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFBertModel: ['cls.predictions.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.dense.bias']
- This IS expected if you are initializing TFBertModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).
All the weights of TFBertModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertModel for predictions w

In [80]:
# Maximum sequence length for BERT
max_length = 128  

# tokenize the text train and test data
train_encodings = tokenizer.batch_encode_plus(train_text.tolist(), max_length=max_length, padding=True, truncation=True)
test_encodings = tokenizer.batch_encode_plus(test_text.tolist(), max_length=max_length, padding=True, truncation=True)

# np.array(encodings['input_ids']) NumPy array contains the tokenized input sequences converted into integer
# np.array(train_encodings['attention_mask']) NumPy array contains the attention mask 
train_inputs = [np.array(train_encodings['input_ids']), np.array(train_encodings['attention_mask'])]
test_inputs = [np.array(test_encodings['input_ids']), np.array(test_encodings['attention_mask'])]

In [81]:
# input_ids and attention_mask are the input layers in the neural networks and input length is 128
input_ids = Input(shape=(max_length,), dtype=tf.int32, name="input_ids")
attention_mask = Input(shape=(max_length,), dtype=tf.int32, name="attention_mask")

# put the input layers inn the pretrained model
outputs = bert_model(input_ids, attention_mask=attention_mask)

# extracts the pooled output from the BERT model output
pooled_output = outputs[1]

# output layer is 2 nodes with softmax as an activation function
output = Dense(2, activation='softmax')(pooled_output)  

# define model input and output
model = Model(inputs=[input_ids, attention_mask], outputs=output)

In [82]:
# compile using ADAM optimizer and loss function is cross entropy
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# fit 1 epoch and batch size equal 16
model.fit(train_inputs, train_labels, epochs=1, batch_size=16)  

4788/4788 [==============================] - 51780s 11s/step - loss: 0.3531 - accuracy: 0.8984


In [83]:
# predict using test data
predictions = model.predict(test_inputs)

# get the higger probabilty to assign the sentence to it from predicted data
predicted_labels = np.argmax(predictions, axis=1)

# get the higger probabilty to assign the sentence to it from actual data
true_labels = np.argmax(test_labels, axis=1)

# print classification report
print(classification_report(true_labels, predicted_labels))

599/599 [==============================] - 2273s 4s/step
              precision    recall  f1-score   support

           0       0.89      1.00      0.94     17128
           1       0.00      0.00      0.00      2021

    accuracy                           0.89     19149
   macro avg       0.45      0.50      0.47     19149
weighted avg       0.80      0.89      0.84     19149



c:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\User\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, mo

#### explaination of the output
- Precision : measure of how many of the predicted positive instances were actually positive
- Recall : measure of how many of the actual positive instances were correctly predicted as positive
- F1-Score : harmonic mean of precision and recall
- Support : number of instances in the true dataset that belong to each class.
- Accuracy : the accuracy of model which is 89 % 
- Macro Avg : These values are the unweighted average of precision, recall, and F1-score for all classes 
- Sample Avg : These values are similar to the micro average but consider each sample equally.